shared/js/background/newtab-tracker-stats.js

```diff
@@ -0,0 +1,348 @@
+ *
+ * Example:
+ *
+ * ```js
+ * const stats = new TrackerStats();
+ * const newtab = new NewTabTrackerStats(stats)
+ *
+ * await newtab.restoreFromStorage();
+ * newtab.register()
+ * ```
+ */
+export class NewTabTrackerStats {
+    /**
+     * The key to use when persisting data into storage
+     */
+    static storageKey = 'trackerStats'
+    /**
+     * The key to use when we don't want to record the company name
+     */
+    static unknownCompanyKey = 'unknown'
+    /**
+     * The prefix used for events. This is used to ensure we only handle events we care about
+     */
+    static eventPrefix = 'newTabPage_'
+
+    /**
+     * @type {Map<string, number> | null}
+     */
+    top100Companies = null
+
+    /**
+     * Internal flag to enable some certain types of logging when developing
+     */
+    _debug = false
+
+    /**
+     * @param {import("./classes/tracker-stats").TrackerStats} stats
+     */
+    constructor (stats) {
+        this.stats = stats
+    }
+
+    /**
+     * Register *all* communications with the extension here.
+     *
+     * The purpose of this method is to co-locate all extension handlers in a single
+     * place for this module.
+     */
+    register () {
+        /**
+         * This listener will redirect the request for tracker-stats.html
+         * on the new tab page to our own HTML file under `web_accessible_resources`
+         */
+        browser.webRequest.onBeforeRequest.addListener(redirectIframeForTrackerStatsMV2,
+            {
+                // todo(Shane): limit this to only the urls we care about
+                urls: ['<all_urls>']
+            },
+            ['blocking']
+        )
+
+        /**
+         * Handle runtime messages sent from the new tab page
+         */
+        browser.runtime.onMessage.addListener((event, sender) => {
+            if (sender.id !== browserWrapper.getExtensionId()) return
+            this.handleIncomingEvent(event)
+        })
+
+        /**
+         * Register an alarm and handle when it fires.
+         * For now we're pruning data every 5 min
+         */
+        const pruneAlarmName = 'pruneNewTabData'
+        browserWrapper.createAlarm(pruneAlarmName, { periodInMinutes: 5 })
+        browser.alarms.onAlarm.addListener(async alarmEvent => {
+            if (alarmEvent.name === pruneAlarmName) {
+                this.handlePruneAlarm()
+            }
+        })
+
+        /**
+         * listen for the 'tracker-blocked' event that is fired from `before-request.es6.js`
+         * when a request is either blocked or a surrogate was used
+         */
+        emitter.on(TrackerBlockedEvent.eventName, (event) => {
+            if (!(event instanceof TrackerBlockedEvent)) return
+            this.record(event.companyDisplayName)
+        })
+
+        /**
+         * Assign the entities from the `tds` data
+         */
+        this.assignTopCompanies(tdsStorage.tds.entities)
```

Comment:
Would it be worth listening for tds updates here too? If a new version of the blocklist comes with new entity names we may need to update the top 100, and maybe also migrate old data.